File: frontend/lib/metascraper/server/diff.ts
import type { AdStatus } from "../types";

export const PROVEN_DAYS = 56;

export interface StoredAd {
  library_id: string;
  niche_id: string;
  page_name: string;
  page_id: string | null;
  primary_text: string;
  headline: string | null;
  description: string | null;
  cta_label: string | null;
  media_type: string;
  platforms: string[];
  started_running_date: string | null;
  ad_library_url: string;
  landing_url: string | null;
  media_url: string | null;
  first_seen: string;
  last_seen: string;
  days_active: number | null;
  weeks_observed: number;
  status: AdStatus;
  hook_category: string | null;
  notes: string | null;
}

export interface CapturePayload {
  captured_date: string;
  country: string;
  ads: Record<string, unknown>[];
  hunted_scope: string[];
}

const VALID_MEDIA = new Set(["image", "video", "carousel", "unknown"]);

function normalizeMedia(v: unknown): string {
  return typeof v === "string" && VALID_MEDIA.has(v) ? v : "unknown";
}

function parseDateMs(value: string | null | undefined): number | null {
  if (!value) return null;
  const d = new Date(value.slice(0, 10));
  return isNaN(d.getTime()) ? null : d.getTime();
}

export function computeDaysActive(startedRunningDate: string | null | undefined, anchor: string): number | null {
  const start = parseDateMs(startedRunningDate);
  const end = parseDateMs(anchor);
  if (start === null || end === null) return null;
  return Math.max(Math.floor((end - start) / 86_400_000), 0);
}

export function isProven(daysActive: number | null | undefined): boolean {
  return daysActive != null && daysActive > PROVEN_DAYS;
}

function scopeContains(scope: Set<string>, ad: StoredAd): boolean {
  if (ad.niche_id && scope.has(ad.niche_id)) return true;
  if (ad.page_id && scope.has(ad.page_id)) return true;
  return false;
}

function adFromRaw(raw: Record<string, unknown>, capturedDate: string): StoredAd {
  const libraryId = String(raw.library_id ?? "");
  return {
    library_id: libraryId,
    niche_id: String(raw.niche_id ?? ""),
    page_name: String(raw.page_name ?? ""),
    page_id: (raw.page_id as string | null) ?? null,
    primary_text: String(raw.primary_text ?? ""),
    headline: (raw.headline as string | null) ?? null,
    description: (raw.description as string | null) ?? null,
    cta_label: (raw.cta_label as string | null) ?? null,
    media_type: normalizeMedia(raw.media_type),
    platforms: Array.isArray(raw.platforms) ? raw.platforms.map(String) : [],
    started_running_date: (raw.started_running_date as string | null) ?? null,
    ad_library_url:
      String(raw.ad_library_url ?? "") ||
      `https://www.facebook.com/ads/library/?id=${libraryId}`,
    landing_url: (raw.landing_url as string | null) ?? null,
    media_url: (raw.media_url as string | null) ?? null,
    first_seen: capturedDate,
    last_seen: capturedDate,
    days_active: null,
    weeks_observed: 0,
    status: "new",
    hook_category: null,
    notes: null,
  };
}

export function mergeCapture(store: StoredAd[], capture: CapturePayload, today: string): StoredAd[] {
  const capturedDate = capture.captured_date;
  const incoming = new Map<string, Record<string, unknown>>();
  for (const raw of capture.ads) {
    const id = String(raw.library_id ?? "");
    if (id) incoming.set(id, raw);
  }

  const byId = new Map(store.map((a) => [a.library_id, a]));

  let scope = new Set(capture.hunted_scope);
  if (scope.size === 0) {
    for (const raw of capture.ads) {
      const n = raw.niche_id;
      if (typeof n === "string" && n) scope.add(n);
    }
  }

  const result: StoredAd[] = [];
  const seenIds = new Set<string>();

  // New + running
  for (const [libraryId, raw] of incoming) {
    seenIds.add(libraryId);
    const existing = byId.get(libraryId);
    const fresh = adFromRaw(raw, capturedDate);

    if (!existing) {
      fresh.status = "new";
      fresh.first_seen = capturedDate;
      fresh.last_seen = capturedDate;
      fresh.weeks_observed = 1;
      fresh.days_active = computeDaysActive(fresh.started_running_date, today);
      result.push(fresh);
    } else {
      const merged: StoredAd = {
        ...existing,
        page_name: fresh.page_name || existing.page_name,
        page_id: fresh.page_id ?? existing.page_id,
        primary_text: fresh.primary_text || existing.primary_text,
        headline: fresh.headline !== null ? fresh.headline : existing.headline,
        description: fresh.description !== null ? fresh.description : existing.description,
        cta_label: fresh.cta_label !== null ? fresh.cta_label : existing.cta_label,
        media_type: raw.media_type ? normalizeMedia(raw.media_type) : existing.media_type,
        platforms: fresh.platforms.length ? fresh.platforms : existing.platforms,
        started_running_date: fresh.started_running_date ?? existing.started_running_date,
        ad_library_url: fresh.ad_library_url || existing.ad_library_url,
        landing_url: fresh.landing_url !== null ? fresh.landing_url : existing.landing_url,
        media_url: fresh.media_url !== null ? fresh.media_url : existing.media_url,
        status: "running",
        last_seen: capturedDate,
        weeks_observed:
          existing.last_seen === capturedDate
            ? existing.weeks_observed
            : existing.weeks_observed + 1,
        hook_category: existing.hook_category,
        notes: existing.notes,
      };
      merged.days_active = computeDaysActive(merged.started_running_date, today);
      result.push(merged);
    }
  }

  // Killed / untouched
  for (const record of store) {
    if (seenIds.has(record.library_id)) continue;
    if (scopeContains(scope, record) && record.status !== "killed") {
      const killed: StoredAd = { ...record, status: "killed" };
      killed.days_active = computeDaysActive(killed.started_running_date, killed.last_seen);
      result.push(killed);
    } else {
      result.push(record);
    }
  }

  return result;
}

export function longevityRanks(store: StoredAd[]): Map<string, number> {
  const byNiche = new Map<string, StoredAd[]>();
  for (const ad of store) {
    if (ad.days_active === null) continue;
    const list = byNiche.get(ad.niche_id) ?? [];
    list.push(ad);
    byNiche.set(ad.niche_id, list);
  }

  const ranks = new Map<string, number>();
  for (const records of byNiche.values()) {
    const ordered = [...records].sort((a, b) => (a.days_active ?? 0) - (b.days_active ?? 0));
    const n = ordered.length;
    ordered.forEach((r, i) => {
      ranks.set(r.library_id, Math.round(((i + 1) / n) * 1000) / 10);
    });
  }
  return ranks;
}

File: backend/metascraper/diff.py
"""Diff & longevity engine (§8). Pure, deterministic, fully unit-testable.

Given the existing AdRecord store and a new CaptureFile, produce the merged
store with new/running/killed status and longevity metrics recomputed. No I/O,
no clock reads except the explicit `today` argument so tests are reproducible.

Key invariants:
  * Match on library_id.
  * Only ads inside the run's hunted_scope can be marked killed — a partial
    hunt must never false-kill untouched niches/competitors.
  * Manual fields (hook_category, notes) survive every merge.
"""

from __future__ import annotations

from dataclasses import replace
from datetime import date

from backend.metascraper.models import AdRecord, CaptureFile, normalize_media_type

PROVEN_DAYS = 56  # 8 weeks — the "proven" longevity threshold.


def _parse_date(value: str | None) -> date | None:
    if not value:
        return None
    try:
        return date.fromisoformat(value[:10])
    except (ValueError, TypeError):
        return None


def compute_days_active(
    started_running_date: str | None,
    anchor: str,
) -> int | None:
    """Whole days between the ad's start date and `anchor` (today if running,
    last_seen if killed). Null-safe: returns None when start date is missing."""
    start = _parse_date(started_running_date)
    end = _parse_date(anchor)
    if start is None or end is None:
        return None
    return max((end - start).days, 0)


def is_proven(days_active: int | None) -> bool:
    return days_active is not None and days_active > PROVEN_DAYS


def _scope_contains(scope: set[str], record: AdRecord) -> bool:
    """Was this record's niche/advertiser actually queried this run? An ad is
    in scope if its niche_id is in the hunted scope, or its page_id was tracked
    as a competitor in scope. niche_id is the primary bound."""
    if record.niche_id and record.niche_id in scope:
        return True
    if record.page_id and record.page_id in scope:
        return True
    return False


def merge_capture(
    store: list[AdRecord],
    capture: CaptureFile,
    today: str,
) -> list[AdRecord]:
    """Merge one capture into the store and return the new store.

    `today` is the ISO date used as the longevity anchor for still-running ads
    (passed explicitly so the engine stays pure and testable).
    """
    captured_date = capture.captured_date
    incoming = {str(a["library_id"]): a for a in capture.ads if a.get("library_id")}
    by_id = {r.library_id: r for r in store}

    # Bound the kill sweep. If the capture didn't declare a scope, fall back to
    # the niches present in the capture's own ads — never the whole store.
    scope: set[str] = set(capture.hunted_scope)
    if not scope:
        scope = {str(a.get("niche_id")) for a in capture.ads if a.get("niche_id")}

    result: list[AdRecord] = []
    seen_ids: set[str] = set()

    # 1. New + running: walk the incoming capture.
    for library_id, payload in incoming.items():
        seen_ids.add(library_id)
        existing = by_id.get(library_id)
        fresh = AdRecord.from_dict(payload)

        if existing is None:
            # NEW — first time we've seen this creative.
            fresh.status = "new"
            fresh.first_seen = captured_date
            fresh.last_seen = captured_date
            fresh.weeks_observed = 1
            fresh.days_active = compute_days_active(fresh.started_running_date, today)
            result.append(fresh)
        else:
            # RUNNING — refresh mutable fields, bump observation counters,
            # preserve manual tags and first_seen.
            merged = replace(
                existing,
                # Refresh capturable fields from the latest payload.
                page_name=fresh.page_name or existing.page_name,
                page_id=fresh.page_id or existing.page_id,
                primary_text=fresh.primary_text or existing.primary_text,
                headline=fresh.headline if fresh.headline is not None else existing.headline,
                description=fresh.description if fresh.description is not None else existing.description,
                cta_label=fresh.cta_label if fresh.cta_label is not None else existing.cta_label,
                media_type=normalize_media_type(payload.get("media_type")) if payload.get("media_type") else existing.media_type,
                platforms=fresh.platforms or existing.platforms,
                started_running_date=fresh.started_running_date or existing.started_running_date,
                ad_library_url=fresh.ad_library_url or existing.ad_library_url,
                landing_url=fresh.landing_url if fresh.landing_url is not None else existing.landing_url,
                media_url=fresh.media_url if fresh.media_url is not None else existing.media_url,
                # Managed counters.
                status="running",
                last_seen=captured_date,
                weeks_observed=existing.weeks_observed + (0 if existing.last_seen == captured_date else 1),
                # Manual fields explicitly preserved.
                hook_category=existing.hook_category,
                notes=existing.notes,
            )
            merged.days_active = compute_days_active(merged.started_running_date, today)
            result.append(merged)

    # 2. Killed / untouched: walk the rest of the store.
    for record in store:
        if record.library_id in seen_ids:
            continue
        if _scope_contains(scope, record) and record.status != "killed":
            # Absent from a run that DID query its niche → killed. Freeze
            # last_seen, anchor days_active to last_seen.
            killed = replace(record, status="killed")
            killed.days_active = compute_days_active(killed.started_running_date, killed.last_seen)
            result.append(killed)
        else:
            # Out of scope this run, or already killed → carry forward untouched.
            result.append(record)

    return result


def longevity_ranks(store: list[AdRecord]) -> dict[str, float]:
    """Percentile rank (0–100) of days_active within each niche. Ads with no
    days_active are excluded from ranking and omitted from the result."""
    ranks: dict[str, float] = {}
    by_niche: dict[str, list[AdRecord]] = {}
    for record in store:
        if record.days_active is None:
            continue
        by_niche.setdefault(record.niche_id, []).append(record)

    for records in by_niche.values():
        ordered = sorted(records, key=lambda r: r.days_active or 0)
        n = len(ordered)
        for index, record in enumerate(ordered):
            # Percentile = share of records at or below this one.
            ranks[record.library_id] = round(((index + 1) / n) * 100, 1)
    return ranks
